Doesn't build against or run on Eclipse 4.10 (2018-12)

If you try to build using Eclipse 4.10 (2018-12) as target platform, you get these failures:

```

[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:1.2.0:compile (default-compile) on project org.testng.eclipse: Compilation failure: Compilation failure: 
[ERROR] /builddir/build/BUILD/testng-eclipse-6.14.0.201802161500/testng-eclipse-plugin/src/main/org/testng/eclipse/ui/conversion/JUnitRewriteCorrectionProposal.java:[7] 
[ERROR] 	import org.eclipse.jdt.internal.corext.codemanipulation.StubUtility;
[ERROR] 	       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ERROR] The import org.eclipse.jdt.internal.corext.codemanipulation.StubUtility cannot be resolved
[ERROR] /builddir/build/BUILD/testng-eclipse-6.14.0.201802161500/testng-eclipse-plugin/src/main/org/testng/eclipse/ui/conversion/JUnitRewriteCorrectionProposal.java:[87] 
[ERROR] 	document.setInitialLineDelimiter(StubUtility.getLineDelimiterUsed(cu));
[ERROR] 	                                 ^^^^^^^^^^^
[ERROR] StubUtility cannot be resolved
[ERROR] 2 problems (2 errors)

```

This is because you used an internal API that changed without warning. Here is the relevant JDT commit: https://git.eclipse.org/c/jdt/eclipse.jdt.ui.git/commit/?id=bc3bc858da1855f31cee9566912859faac05819e

This should be fixed before this plugin can work on the latest 2018-12 release of Eclipse

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -14,6 +14,7 @@ Supported Metrics:
 * PR #401: support TestNG 7.0
 * Fixed #348: String index out of range: -2. (@timrsfo & Nick Tan)
 * Fixed #407: eclipse plugin generates invalid java for toString & hashCode (@dabraham02124 & Nick Tan)
+* Fixed #408: Doesn't build against or run on Eclipse 4.10 (2018-12)
 
 ## 6.14
 

--- a/pom.xml
+++ b/pom.xml
@@ -53,17 +53,7 @@
                     <environment>
                       <os>win32</os>
                       <ws>win32</ws>
-                      <arch>x86</arch>
-                    </environment>
-                    <environment>
-                      <os>win32</os>
-                      <ws>win32</ws>
                       <arch>x86_64</arch>
-                    </environment>
-                    <environment>
-                      <os>linux</os>
-                      <ws>gtk</ws>
-                      <arch>x86</arch>
                     </environment>
                     <environment>
                       <os>linux</os>

--- a/testng-eclipse-plugin/src/main/org/testng/eclipse/ui/conversion/JUnitRewriteCorrectionProposal.java
+++ b/testng-eclipse-plugin/src/main/org/testng/eclipse/ui/conversion/JUnitRewriteCorrectionProposal.java
@@ -2,9 +2,17 @@ package org.testng.eclipse.ui.conversion;
 
 import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.core.runtime.preferences.IScopeContext;
+import org.eclipse.core.runtime.preferences.InstanceScope;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.ProjectScope;
 import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IOpenable;
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.JavaModelException;
 import org.eclipse.jdt.core.dom.rewrite.ASTRewrite;
-import org.eclipse.jdt.internal.corext.codemanipulation.StubUtility;
 import org.eclipse.jdt.ui.text.java.IJavaCompletionProposal;
 import org.eclipse.jface.text.BadLocationException;
 import org.eclipse.jface.text.Document;
@@ -84,7 +92,7 @@ public class JUnitRewriteCorrectionProposal implements IJavaCompletionProposal
 //      }
 //    }
     Document document= new Document(cu.getSource());
-    document.setInitialLineDelimiter(StubUtility.getLineDelimiterUsed(cu));
+    document.setInitialLineDelimiter(getLineDelimiterUsed(cu));
     change= new DocumentChange(name, document);
 //    } else {
 //      Document doc = new Document(m_cu.getSource());
@@ -99,6 +107,45 @@ public class JUnitRewriteCorrectionProposal implements IJavaCompletionProposal
 //    IDocument document= change.getCurrentDocument(new NullProgressMonitor());
     addEdits(change.getCurrentDocument(new NullProgressMonitor()), rootEdit);
     return change;
+  }
+
+  /* This method is based on the implementation from org.eclipse.jdt.internal.core.manipulation.StubUtility */
+  private static String getLineDelimiterUsed(ICompilationUnit elem) {
+    IOpenable openable= elem.getOpenable();
+    if (openable instanceof ITypeRoot) {
+      try {
+        return openable.findRecommendedLineSeparator();
+      } catch (JavaModelException exception) {
+        // Use project setting
+      }
+    }
+    IJavaProject javaProject= elem.getJavaProject();
+    IProject project= null;
+    if (javaProject != null) {
+      project= javaProject.getProject();
+    }
+    String lineDelimiter= getLineDelimiterPreference(project);
+    if (lineDelimiter != null) {
+      return lineDelimiter;
+    }
+    return System.getProperty("line.separator", "\n"); //$NON-NLS-1$ //$NON-NLS-2$
+  }
+
+  /* This method is based on the implementation from org.eclipse.jdt.internal.core.manipulation.StubUtility */
+  private static String getLineDelimiterPreference(IProject project) {
+    IScopeContext[] scopeContext;
+    if (project != null) {
+      // project preference
+      scopeContext= new IScopeContext[] { new ProjectScope(project) };
+      String lineDelimiter= Platform.getPreferencesService().getString(Platform.PI_RUNTIME, Platform.PREF_LINE_SEPARATOR, null, scopeContext);
+      if (lineDelimiter != null) {
+        return lineDelimiter;
+      }
+    }
+    // workspace preference
+    scopeContext= new IScopeContext[] { InstanceScope.INSTANCE };
+    String platformDefault= System.getProperty("line.separator", "\n"); //$NON-NLS-1$ //$NON-NLS-2$
+    return Platform.getPreferencesService().getString(Platform.PI_RUNTIME, Platform.PREF_LINE_SEPARATOR, platformDefault, scopeContext);
   }
 
   protected void addEdits(IDocument document, TextEdit editRoot) throws CoreException {

